### ISSUE TITLE
ActivityId id is empty in async context in ASP.NET Web API

### ISSUE THREAD

**Issue by SirBisgaard** [NLog/NLog.Web#386]
Hi I am using the activityid in our logging for getting a correlation ID, and i followed the guide that is provided on GitHub. That states i only need to add a method in global asax, where we set the activity id.

All my api's calls are mostly async methods and some times the activity id is empty?

**NLog version**: 4.6.2
**NLog.Config version**: 4.6.2
**NLog.Schema version**: 4.6.2
**NLog.Web version**: 4.8.1

**Platform**:  .Net 4.5

**Current NLog config** (xml or C#, if relevant)
```xml
  <nlog throwConfigExceptions="true">
    <targets>
      <target name="debugger" type="Debugger" layout="${logger}::${message}" />
      <target name="database" type="Database" dbProvider="sqlserver" connectionStringName="Nlog">
        <commandText>
          INSERT INTO [dbo].[Log] (MachineName, Service, Logged, CorrelationId, Level, Message, Logger, Callsite, Exception) VALUES (@MachineName, 'My service name', @Logged, @CorrelationId, @Level, @Message, @Logger, @Callsite, @Exception);
        </commandText>
        <parameter name="@MachineName" layout="${machinename}" />
        <parameter name="@Logged" layout="${date}" />
        <parameter name="@CorrelationId" layout="${activityid}" />
        <parameter name="@Level" layout="${level}" />
        <parameter name="@Message" layout="${message}" />
        <parameter name="@Logger" layout="${logger}" />
        <parameter name="@Callsite" layout="${callsite}" />
        <parameter name="@Exception" layout="${exception:tostring}" />
      </target>
    </targets>
    <rules>
      <logger name="*" minlevel="Trace" writeTo="debugger" />
      <logger name="*" minlevel="Trace" writeTo="database" />
    </rules>
  </nlog>
```

I was reading about this specific problem with async calls and the global asax not being called and I found a solution to this.

I created my own DelegatingHandler that intercepts all messages sent to the service. Where i get a correlation Id and manually sets the ActivityId. 
```c#
using System;
using System.Diagnostics;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;

public class CorrelationHandler : DelegatingHandler
    {
        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            Guid correlationId = request.GetCorrelationId();
            Trace.CorrelationManager.ActivityId = correlationId;

            return await base.SendAsync(request, cancellationToken);
        }
    }
```

Finally I registered my new handler in the WebApiConfig.cs file like so: 
```c#

    public static class WebApiConfig
    {
        public static void Register(HttpConfiguration config)
        {
            ...
            // Web API message handlers
            config.MessageHandlers.Add(new CorrelationHandler());
            ...
        }
    }
```

**Comment by 304NotModified** [comment]
Hi! 

Thanks for the feedback. This is more an idea how to implement this in NLog itself?

**Comment by SirBisgaard** [comment]
Hi @304NotModified I will take a look at the NLog.Web package and see how, but i would recommend to init the DelegatingHandler when using the activityid in the config or as a extension method on HttpConfiguration, that is used in the WebApiConfig class.

**Comment by 304NotModified** [comment]
any luck on this?

**Comment by SirBisgaard** [comment]
Sorry for the very slow answer.... :( (Notification settings updated*)

The implementation I made is running in our production and working perfectly! 👍 Its running on multiple services that is logging a ton of data, with a large tree of logging that is now easy to sort by request. 
If I need to provide more examples just say so. 

Not related to this, I'm working on a way to share correlation id's between services so we (my work) are able to sort a request through our whole service stack. Is this something that could be used in NLog in any way?

**Comment by 304NotModified** [comment]
Oops, sorry for the delay

> Not related to this, I'm working on a way to share correlation id's between services so we (my work) are able to sort a request through our whole service stack.

Is that in the same process? 

> The implementation I made is running in our production and working perfectly!

Could you please send a PR?

**Comment by 304NotModified** [comment]
polite bump @SirBisgaard

**Comment by SirBisgaard** [comment]
@304NotModified I'm very sorry for my slow work.... :(
I have set some time off to get it done this weekend. Work is taking all my time at the moment.

**Comment by 304NotModified** [comment]
No problem! Would be great to have this issue fixed!

**Comment by SirBisgaard** [comment]
I am having troubles with the implementation. So I will describe what needs changing. :-) 

In the AspNetTraceIdentifierLayoutRenderer.cs. Do we need to add the following section, but only for .NET 4.6 applications. I am not sure if 4.5 or 4.7 is supported.
I cannot find the method in the Microsoft docs but it works in our web API application, running 4.6.1.
```c# 
private string LookupTraceIdentifier(HttpContext httpContext)
{
      var requestMessage = httpContext.Items["MS_HttpRequestMessage"] as System.Net.Http.HttpRequestMessage;
      var correlationId = requestMessage.GetCorrelationId();
      return correlationId;
}
```
I hope this gives some value. :-)

**Comment by 304NotModified** [comment]
This is one of the docs?

https://docs.microsoft.com/en-us/previous-versions/aspnet/mt174862(v=vs.118)

**Comment by 304NotModified** [comment]
For aspnet core it's https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.httpcontext.traceidentifier?view=aspnetcore-2.2

**Comment by SirBisgaard** [comment]
That's the method. :) 
It's browsable from Google but not from navigation or searching on the documentation.

**Comment by 304NotModified** [comment]
Have you tried `${aspnet-traceidentifier}` ?

**Comment by SirBisgaard** [comment]
Yes it does not work in .NET 4.6. the same with activity id

**Comment by snakefoot** [comment]
Sounds like a duplicate of  #328 (Change requires an extra dependency on `Windows.Web.Http.dll`)

Would it not be possible to activate ETW-Trace on .NET 4.6 and get the correlationid? See also https://docs.microsoft.com/en-us/iis/configuration/system.webServer/httpTracing/ (And maybe #330)

**Comment by 304NotModified** [comment]
@snakefoot  it this then fixed by #330 ?

**Comment by snakefoot** [comment]
> @304NotModified it this then fixed by #330 ?

No idea. Never got a response from @SirBisgaard or @Giorgi whether enabling ETW-Tracing on ASP.NET worked as a solution.

Could probably create a funky layout-renderer for NLog.Web that extracts httpContext.Items["MS_HttpRequestMessage"] and then with reflections calls GetCorrelationId() on the returned object. Then they would get precisely what they want.

**Comment by 304NotModified** [comment]
I will assume this as fixed. Otherwise will reopen
